lightrag/lightrag/components/model_client/ollama_client.py

```diff
@@ -0,0 +1,153 @@
+            embeddings = Embedding(embedding=response['embedding'])
+            return EmbedderOutput(data=[embeddings])
+        except Exception as e:
+            log.error(f"Error parsing the embedding response: {e}")
+            return EmbedderOutput(data=[], error=str(e), raw_response=response)
+
+    def convert_inputs_to_api_kwargs(
+        self,
+        input: Optional[Any] = None, 
+        model_kwargs: Dict = {},
+        model_type: ModelType = ModelType.UNDEFINED,
+    ) -> Dict:
+        r"""
+        For LLM, expect model_kwargs to have the following keys:
+         model: str,
+         messages: str,
+
+        For EMBEDDER, expect model_kwargs to have the following keys:
+         model: str,
+         prompt: str,
+        """
+        final_model_kwargs = model_kwargs.copy()
+        if model_type == ModelType.EMBEDDER:
+            # make sure input is a string
+            if input is not None and input != "":
+                final_model_kwargs["prompt"] = input
+                return final_model_kwargs
+            else:
+                raise ValueError("input must be text")
+        if model_type == ModelType.LLM:
+            messages: List[Dict[str, str]] = []
+            if input is not None and input != "":
+                messages.append({"role": "user", "content": input})
+            else:
+                raise ValueError("Input must be text")
+            final_model_kwargs["messages"] = messages
+        else:
+            raise ValueError(f"model_type {model_type} is not supported")
+        return final_model_kwargs
+
+    @backoff.on_exception(
+        backoff.expo,
+        (RequestError,),
+        max_time=5,
+    )
+    def call(self, api_kwargs: Dict = {}, model_type: ModelType = ModelType.UNDEFINED):
+        """
+        kwargs is the combined input and model_kwargs
+        """
+        if "model" not in api_kwargs: 
+            raise ValueError("model must be specified")
+        log.info(f"api_kwargs: {api_kwargs}")
+        if model_type == ModelType.EMBEDDER:
+            return self.sync_client.embeddings(**api_kwargs)
+        if model_type == ModelType.LLM:
+            return self.sync_client.chat(**api_kwargs)
+        else:
+            raise ValueError(f"model_type {model_type} is not supported")
+
+    @backoff.on_exception(
+        backoff.expo,
+        (RequestError),
+        max_time=5,
+    )
+    async def acall(
+        self, api_kwargs: Dict = {}, model_type: ModelType = ModelType.UNDEFINED
+    ):
+        """
+        kwargs is the combined input and model_kwargs
+        """
+        if self.async_client is None:
+            self.init_async_client()
+        if "model" not in api_kwargs: 
+            raise ValueError("model must be specified")
+        if model_type == ModelType.EMBEDDER:
+            return await self.async_client.embeddings(**api_kwargs)
+        if model_type == ModelType.LLM:
+            return await self.async_client.chat(**api_kwargs)
```

Comment:
another thing we need to think about is why using chat instead of `generate`, the difference between them

Reply:
Generate accepts `prompt: str` as the  parameter, while chat accepts `messages: Optional[Sequence[Message]]`.
When it comes to the underlying difference, it's not as clear.
https://github.com/ollama/ollama/issues/2774

It probably makes more sense to go with generate, since the core of LightRAG contains Prompts which are designed to give the developer the formatting flexibility.